modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,596 @@
+/*
+ * PROJECT:         ReactOS kernel-mode tests
+ * LICENSE:         GPLv2+ - See COPYING in the top level directory
+ * PURPOSE:         Kernel-Mode Test Suite - ObInsertObject test
+ * PROGRAMMER:      Gleb Surikov <glebs.surikovs@gmail.com>
+ */
+
+#include <kmt_test.h>
+
+#define NDEBUG
+#include <debug.h>
+
+typedef struct DUMMY_TYPE
+{
+    ULONG Data;
+} DUMMY_TYPE, *PDUMMY_TYPE;
+
+static HANDLE DirectoryHandle;
+static UNICODE_STRING DirectoryName;
+
+static POBJECT_TYPE DummyType;
+static UNICODE_STRING DummyTypeName;
+static OBJECT_TYPE_INITIALIZER ObjectTypeInitializer;
+static OBJECT_ATTRIBUTES DummyObjectAttributes;
+
+/* This counter is incremented by 1 in the delete callback procedure.
+   This way we know that the object has been deleted and this is used
+   to test the dereferencing and deletion of the object in case ObInsertObject fails */
+static ULONG ObjectDeletionCounter = 0;
```

Comment:
Suggestion, without such variable and callback: add an extra reference, trigger failure, check internal reference count decreased, remove the added reference.

Reply:
What I like is that by leaving 0 references we can check the deletion of the object itself and the execution of the conditional deletion of the object as part of retention control